QuizApp/MVP/Model/Repository/QuizRepository.swift

```diff
@@ -8,13 +8,98 @@
+                if !self.quizNetworkSource.checkReachability(){
+                    let quizzes:[Quiz] = self.quizDatabaseSourece.fetchQuizzesFromCoreData(filter: FilterSettings(searchText: nil))
+                    if quizzes.isEmpty{
+                        presenter.delegate.showNoQuizzes()
+                    }else{
+                        presenter.proccesAndShowQuizzes(allQuizzes: quizzes)
+                    }
+                }else {
+                // network
+                    self.quizNetworkSource.getQuizzes(){
+                    (result: Result<QuizzesResponse, RequestError>) in
+                        switch result {
+                            case .failure(let error):
+                                // dont show error, offline mode is here
+                                // self.showError(error: error,presenter: presenter)
+                                print(error)
+                            case .success(let value):
+                                if value.quizzes.count == 0 {
+                                    self.quizDatabaseSourece.saveNewQuizzes(value.quizzes, imagesDictionary: self.imagesDictionary)
+                                    presenter.delegate.showNoQuizzes()
+                                }else{
+                                    // download images first time and on new quizzes on server
+                                    if self.imagesDictionary.isEmpty || self.imagesDictionary.keys.count != value.quizzes.count {
+                                        self.imagesDictionary = self.getQuizzesImages(quizzes: value.quizzes, connection: true)
+                                    }
+                                    self.quizDatabaseSourece.saveNewQuizzes(value.quizzes, imagesDictionary: self.imagesDictionary)
+                                    // getting quizzes from database to get downloaded images
+                                    let quizzesWithImages = self.quizDatabaseSourece.fetchQuizzesFromCoreData(filter: FilterSettings(searchText: nil))
+                                    presenter.proccesAndShowQuizzes(allQuizzes: quizzesWithImages)
+                                }
+                        }
+                    }
+            }
+        }
+    }
+    
+    func getFilteredQuizzes(presenter:QuizzesPresenter, filterText:String?){
+        DispatchQueue.global(qos: .userInitiated).async {
+            if filterText == nil || filterText == "" {
+                presenter.delegate.showNoQuizzes()
+            }else{
+                let quizzes:[Quiz] = self.quizDatabaseSourece.fetchQuizzesFromCoreData(filter: FilterSettings(searchText: filterText))
+                if quizzes.isEmpty{
+                    presenter.delegate.showNoQuizzes()
+                }else{
+                    presenter.proccesAndShowQuizzes(allQuizzes: quizzes)
+                }
+            }
+        }
+    }
+    
+    
+    func showError(error:RequestError,presenter:QuizzesPresenter){
```

Comment:
This does not belong in QuizRepository. Your QuizzesPresenter has a reference to this class, so ideally you don't want to have a two way reference.